backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java

```diff
@@ -0,0 +1,178 @@
+
+    } else {
+      // Create the vision manager object and get the labels as JSON
+      VisionManager visionManager = new VisionManagerImpl(bytes);
+      Text labelsJsonArrayString = new Text(visionManager.getLabelsAsJson());
+
+      // Add the input to datastore
+      Entity analyzedImageEntity = new Entity("analyzed-image");
+      analyzedImageEntity.setProperty("imageUrl", imageUrl);
+      analyzedImageEntity.setProperty("labelsJsonArray", labelsJsonArray);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(analyzedImageEntity);
+
+      // Redirect back to the HTML page.
+      response.sendRedirect("/vision-upload-prototype/vision-demo.html#image-upload");
+    }
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
+  }
+
+  /** Returns image byte data from BlobKey of image stored with Blobstore. */
+  private byte[] getBlobBytes(HttpServletRequest request, String formInputElementName)
+      throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
```

Comment:
Potentially nit: Consider throwing a detailed exception instead of returning null for these cases.